src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,192 @@
+//! Paths and Unix shells
+//!
+//! MacOS, Linux, FreeBSD, and many other OS model their design on Unix,
+//! so handling them is relatively consistent. But only relatively.
+//! POSIX postdates Unix by 20 years, and each "Unix-like" shell develops
+//! unique quirks over time.
+//!
+//!
+//! Windowing Managers, Desktop Environments, GUI Terminals, and PATHs
+//!
+//! Duplicating paths in PATH can cause performance issues when the OS searches
+//! the same place multiple times. Traditionally, Unix configurations have
+//! resolved this by setting up PATHs in the shell's login profile.
+//!
+//! This has its own issues. Login profiles are only intended to run once, but
+//! changing the PATH is common enough that people may run it twice. Desktop
+//! environments often choose to NOT start login shells in GUI terminals. Thus,
+//! a trend has emerged to place PATH updates in other run-commands (rc) files,
+//! leaving Rustup with few assumptions to build on for fulfilling its promise
+//! to set up PATH appropriately.
+//!
+//! Rustup addresses this by:
+//! 1) using a shell script that updates PATH if the path is not in PATH
+//! 2) sourcing this script in any known and appropriate rc file
+
+use super::canonical_cargo_home;
+use super::*;
+use crate::process;
+use std::path::PathBuf;
+
+pub type Shell = Box<dyn UnixShell>;
+
+#[derive(Debug, PartialEq)]
+pub struct ShellScript {
+    content: &'static str,
+    name: &'static str,
+}
+
+impl ShellScript {
+    pub fn write(&self) -> Result<()> {
+        let cargo_bin = format!("{}/bin", canonical_cargo_home()?);
+        let env_name = utils::cargo_home()?.join(self.name);
+        let env_file = self.content.replace("{cargo_bin}", &cargo_bin);
+        utils::write_file(self.name, &env_name, &env_file)?;
+        Ok(())
+    }
+}
+
+#[allow(dead_code)] // For some reason.
+const POSIX_ENV: &str = include_str!("env");
+
+macro_rules! support_shells {
+    ( $($shell:ident,)* ) => {
+        fn enumerate_shells() -> Vec<Shell> {
+            vec![$( Box::new($shell), )*]
+        }
+    }
+}
+
+// TODO: Tcsh (BSD)
+// TODO?: Make a decision on Ion Shell, Power Shell, Nushell
+// Cross-platform non-POSIX shells have not been assessed for integration yet
+support_shells! {
+    Posix,
+    Bash,
+    Zsh,
+}
+
+pub fn get_available_shells() -> impl Iterator<Item = Shell> {
+    enumerate_shells().into_iter().filter(|sh| sh.does_exist())
+}
+
+pub trait UnixShell {
+    // Detects if a shell "exists". Users have multiple shells, so an "eager"
+    // heuristic should be used, assuming shells exist if any traces do.
+    fn does_exist(&self) -> bool;
+
+    // Gives all rcfiles of a given shell that rustup is concerned with.
+    // Used primarily in checking rcfiles for cleanup.
+    fn rcfiles(&self) -> Vec<PathBuf>;
```

Comment:
See below - but I think rcfiles vs update_rcs is an unneeded distinction, because we can be pretty certain we won't be `pure` as far as updating bash's profile files is concerned, because of the intersection with the posix shell definition.

Reply:
Mm, the evolution of the zsh implementation suggests it is going to be a concern anyways.

Reply:
Looking through this, I cannot see rcfiles() being called on the trait; only from within trait implementations. All I'm suggesting is that it be removed from the trait unless-and-until something that is calling methods on the trait needs to call it.